NIFI-5952 [WIP] Fixing RAW S2S illegal blocking mode on Java 11

ijokarumawak opened the pull request:
Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

Another approach to address RAW S2S illegal blocking mode error on Java 11 reported with NIFI-5952. 
While #3265 refactors NOT to use NIO (thread/conn instead), this PR tries implementing NIO correctly.

Performance test confirmed NIO S2S can support 2,000 concurrent connections while thread/conn can only accept few hundreds.

This PR is still in draft state as it has SocketTimeout issue at secured NiFi environment in some situations. Still investigating that issue, but sending a PR for early code review.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically `master`)?

- [x] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [x] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

markap14 commented on nifi-commons/nifi-site-to-site-client/src/main/java/org/apache/nifi/remote/codec/StandardFlowFileCodec.java:
```diff
@@ -56,13 +62,32 @@ public void encode(final DataPacket dataPacket, final OutputStream encodedOut) t
+        final byte[] buffer = new byte[8192];
+        long totalRead = 0;
+        // Use 'available' so that it doesn't block
+        for (int available; totalRead < length && (available = source.available()) > 0;) {
```
We cannot use InputStream.available() for this purpose. See https://docs.oracle.com/javase/7/docs/api/java/io/InputStream.html#available() for an explanation of the method. It is an "estimate" of the number of bytes, but a great number of InputStreams simply return `0`, which makes the result unreliable. In such a case, this will immediately return without copying any bytes.

ijokarumawak replied:
Thank you for catching this. I think this is a leftover while I was trying different things. Reverted changes to StandardFlowFileCodec.

markap14 commented:
@ijokarumawak I tried testing this out locally. I created a 2-node cluster, secured. I created a single "Public" InputPort and gave it permissions. Then generated some data and attempted to send via RPG with RAW protocol. However, nothing would send over Site-to-Site. I disabled the RPG and enabled again. 925 of the 20,000 FlowFiles (each 1 KB) were sent, but then it stopped sending. Tried disabling and re-enabling again. Nothing else was transferred. After a while, started seeing Read Timeouts and SocketException: Connection Reset.

markap14 commented:
Also of note, CPU utilization is very high. One of the instances is at 200% CPU utilization, the other at 300% utilization. The only thing running is the RPG and the Public Input Port.

markap14 commented:
The two Timer-Driven threads that are active:

```
"Timer-Driven Process Thread-5" Id=114 RUNNABLE
        at org.apache.nifi.stream.io.ByteCountingInputStream.read(ByteCountingInputStream.java:51)
        at org.apache.nifi.remote.io.InterruptableInputStream.read(InterruptableInputStream.java:57)
        at java.io.DataInputStream.read(DataInputStream.java:149)
        at java.util.zip.CheckedInputStream.read(CheckedInputStream.java:82)
        at java.io.DataInputStream.readFully(DataInputStream.java:195)
        at java.io.DataInputStream.readLong(DataInputStream.java:416)
        at org.apache.nifi.remote.codec.StandardFlowFileCodec.decode(StandardFlowFileCodec.java:116)
        at org.apache.nifi.remote.protocol.AbstractFlowFileServerProtocol.receiveFlowFiles(AbstractFlowFileServerProtocol.java:442)
        at org.apache.nifi.remote.StandardPublicPort.receiveFlowFiles(StandardPublicPort.java:268)
        at org.apache.nifi.remote.StandardPublicPort.onTrigger(StandardPublicPort.java:245)
        at org.apache.nifi.remote.StandardPublicPort.onTrigger(StandardPublicPort.java:180)
        at org.apache.nifi.controller.tasks.ConnectableTask.invoke(ConnectableTask.java:213)
        at org.apache.nifi.controller.scheduling.TimerDrivenSchedulingAgent$1.run(TimerDrivenSchedulingAgent.java:117)
        at org.apache.nifi.engine.FlowEngine$2.run(FlowEngine.java:110)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
        Number of Locked Synchronizers: 1
        - java.util.concurrent.ThreadPoolExecutor$Worker@7c2acb04
```
```
"Timer-Driven Process Thread-6" Id=117 RUNNABLE  (in native code)
        at java.net.SocketInputStream.socketRead0(Native Method)
        at java.net.SocketInputStream.socketRead(SocketInputStream.java:116)
        at java.net.SocketInputStream.read(SocketInputStream.java:170)
        at java.net.SocketInputStream.read(SocketInputStream.java:141)
        at sun.security.ssl.InputRecord.readFully(InputRecord.java:465)
        at sun.security.ssl.InputRecord.read(InputRecord.java:503)
        at sun.security.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:973)
        - waiting on java.lang.Object@fb69dd5
        at sun.security.ssl.SSLSocketImpl.readDataRecord(SSLSocketImpl.java:930)
        at sun.security.ssl.AppInputStream.read(AppInputStream.java:105)
        - waiting on sun.security.ssl.AppInputStream@74a19cc3
        at org.apache.nifi.stream.io.ByteCountingInputStream.read(ByteCountingInputStream.java:51)
        at java.io.BufferedInputStream.fill(BufferedInputStream.java:246)
        at java.io.BufferedInputStream.read(BufferedInputStream.java:265)
        - waiting on java.io.BufferedInputStream@471e5956
        at org.apache.nifi.remote.io.InterruptableInputStream.read(InterruptableInputStream.java:39)
        at java.io.FilterInputStream.read(FilterInputStream.java:83)
        at org.apache.nifi.remote.protocol.ResponseCode.readCode(ResponseCode.java:117)
        at org.apache.nifi.remote.protocol.Response.read(Response.java:43)
        at org.apache.nifi.remote.protocol.socket.SocketClientTransaction.readTransactionResponse(SocketClientTransaction.java:82)
        at org.apache.nifi.remote.AbstractTransaction.confirm(AbstractTransaction.java:239)
        at org.apache.nifi.remote.client.socket.SocketClient$2.confirm(SocketClient.java:147)
        at org.apache.nifi.remote.StandardRemoteGroupPort.transferFlowFiles(StandardRemoteGroupPort.java:369)
        at org.apache.nifi.remote.StandardRemoteGroupPort.onTrigger(StandardRemoteGroupPort.java:284)
        at org.apache.nifi.controller.AbstractPort.onTrigger(AbstractPort.java:245)
        at org.apache.nifi.controller.tasks.ConnectableTask.invoke(ConnectableTask.java:213)
        at org.apache.nifi.controller.scheduling.TimerDrivenSchedulingAgent$1.run(TimerDrivenSchedulingAgent.java:117)
        at org.apache.nifi.engine.FlowEngine$2.run(FlowEngine.java:110)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
        Number of Locked Synchronizers: 1
        - java.util.concurrent.ThreadPoolExecutor$Worker@438eba94
```

ijokarumawak commented:
@markap14 Thank you for taking a look! Fixed few broken part when I rebased with the master. But it still isn't working properly at a secured cluster environment. Keep investigating..

markap14 commented:
@ijokarumawak I merged in PR #3265, which is an alternative to this PR. Perhaps it makes sense to close this PR for now, and if you're able to get the issues resolved for this NIO-based implementation we can create a new PR against a new Jira?

ijokarumawak commented:
Thanks @markap14 for merging #3265. I agree with you, closing this one.
